core/trino-main/src/main/java/io/trino/operator/DeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,358 @@
+import io.airlift.units.DataSize;
+import io.trino.execution.TaskId;
+import io.trino.execution.buffer.SerializedPage;
+import io.trino.spi.TrinoException;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executor;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.util.concurrent.Futures.nonCancellationPropagating;
+import static io.trino.operator.RetryPolicy.NONE;
+import static io.trino.operator.RetryPolicy.QUERY;
+import static io.trino.operator.RetryPolicy.TASK;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+public class DeduplicationExchangeClientBuffer
+        implements ExchangeClientBuffer
+{
+    private final Executor executor;
+    private final long bufferCapacityInBytes;
+    private final RetryPolicy retryPolicy;
+
+    private final SettableFuture<Void> blocked = SettableFuture.create();
+    @GuardedBy("this")
+    private final Set<TaskId> allTasks = new HashSet<>();
+    @GuardedBy("this")
+    private boolean noMoreTasks;
+    @GuardedBy("this")
+    private final Set<TaskId> successfulTasks = new HashSet<>();
+    @GuardedBy("this")
+    private final Map<TaskId, Throwable> failedTasks = new HashMap<>();
+    @GuardedBy("this")
+    private boolean inputFinished;
+    @GuardedBy("this")
+    private Throwable failure;
+
+    @GuardedBy("this")
+    private final ListMultimap<TaskId, SerializedPage> pageBuffer = LinkedListMultimap.create();
+    @GuardedBy("this")
+    private Iterator<SerializedPage> pagesIterator;
+    @GuardedBy("this")
+    private volatile long bufferRetainedSizeInBytes;
+    @GuardedBy("this")
+    private volatile long maxBufferRetainedSizeInBytes;
+    @GuardedBy("this")
+    private int maxAttemptId;
+
+    @GuardedBy("this")
+    private boolean closed;
+
+    public DeduplicationExchangeClientBuffer(Executor executor, DataSize bufferCapacity, RetryPolicy retryPolicy)
+    {
+        this.executor = requireNonNull(executor, "executor is null");
+        this.bufferCapacityInBytes = requireNonNull(bufferCapacity, "bufferCapacity is null").toBytes();
+        requireNonNull(retryPolicy, "retryPolicy is null");
+        checkArgument(retryPolicy != NONE, "retryPolicy is not expected to be NONE");
+        this.retryPolicy = retryPolicy;
+    }
+
+    @Override
+    public ListenableFuture<Void> isBlocked()
+    {
+        return nonCancellationPropagating(blocked);
```

Comment:
The assumption here is that clients are supposed to wait for the `isBlocked` future before performing any further operations. Thus in theory the number of listeners shouldn't exceed the number of threads interacting with the buffer (that in practice should not be high). In the old implementation `ExchangeClient` used to create a new future for every `isBlocked` call what seemed to be even worse than attaching a listener.

To avoid creating extra listeners there are several approaches we can take:

1. Don't wrap the future in `nonCancellationPropagating` and assume clients never cancel. This seem to be rather fragile and I'm not sure if that's in fact always true today.
2. Cache the `nonCancellationPropagating` future as a field. This will add an extra complexity as both fields will have to be maintained in a consistent way. 
3. Implement "nonCancellingSettableFuture". That seem to be the most compelling one, but it will also introduce some amount of extra complexity.

Reply:
Ok, let's leave it as is for now.